api/lib/query.py

```diff
@@ -0,0 +1,156 @@
+from .db.database import GameArchive, init as init_db
+from .db.database import db_session, Participant, Game, Move
+from sqlalchemy import or_
+
+from .constants import STOCKFISH_INVITEE_ID
+
+def init(dbURL):
+    init_db(dbURL)
+
+def get_participant(discordID):
+    # discordID could either be the discord user ID or 
+    # the GuildMemberID
+    db = db_session()
+    return _get_participant_with_session(db, discordID)
+
+def _get_participant_with_session(db, discordID):
+    return db.query(Participant).\
+        filter(or_(
+            Participant.discord_user_id == discordID,
+            Participant.discord_guild_id == discordID
+        )).\
+        order_by(Participant.id.desc()).\
+        first()
+
+def get_participant_from_id(db_id):
+    db = db_session()
+    return db.query(Participant).filter_by(id=db_id).first()
+
+def create_participant(user_id, guild_id):
+    db = db_session()
+    try:
+        p = Participant(discord_user_id = user_id, discord_guild_id = guild_id)
+        db.add(p)
+        db.commit()
+        return True
+    except:
+        db.rollback()
+        return False
+
+def update_participant(user_id, guild_id):
+    db = db_session()
+    try:
+        user = _get_participant_with_session(db, user_id)
+        user.guild_id = guild_id
+        db.commit()
+        return True
+    except:
+        db.rollback()
+        return False
+        
+def get_recent_game(participant):
+    db = db_session()
+    return db.query(Game).\
+        filter(or_(
+            Game.author_id == participant.id,
+            Game.invitee_id == participant.id
+        )).\
+        order_by(Game.last_updated.desc()).\
+        first()
+
+def get_solo_game(participant):
+    db = db_session()
+    return db.query(Game).\
+        filter_by(author_id=participant.id).\
+        filter_by(invitee_id=STOCKFISH_INVITEE_ID).\
+        first()
+
+def get_pvp_game(participant, invitee):
+    db = db_session()
+    return db.query(Game).\
+        filter_by(author_id=participant.id).\
+        filter_by(invitee_id=invitee.id).\
+        first()
+
+def get_moves_string(game):
```

Comment:
>mentioned this already in a previous comment but get_moves_string should ideally be O(1) since players making moves is the single most common user scenario for a chess bot

I understand the concept. Make the data a fixed size in memory so that way the data stays sequential. The O(N) operation mentioned being reconstructing the data when read from the database.

However, I think this could be a preoptimization, since we have to feed our moves through STOCKFISH per verification, most likely the STOCKFISH AI will have to iterate through the dataset anyway. This means that the operation adding moves will always be O(N) regardless of the data structure.

In addition, by making the column data type for the data fixed, you assume how many moves there are in a game of chess. Ideally the max is 5899 moves, but unless we want to start stripping games of infinite moves, your database queries will fail (potentially crashing apps) if we stick to that rule. This form of data processing is more robust and can be easily fixable (i.e. through a future retention policy) rather than sticking to TEXT based blob columns.

Finally, we also allow for smaller games to have less of an impact on the resolution of games. By giving games each a fixed amount of memory does allow for games to be quickly created and abandoned while leaving the devs/app at a disadvantage due to disk/memory fragmentation. Using the index will provide enough speed that runtime complexity should not be too stringent on this application.

Reply:
I switched back to TEXT columns anyways. I kept the GameArchive table as those the new Database locks will allow users to only have one game open per member/user.